Generalize sizes and strides policy on _make_wrapper_subclass

suo opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #78646

Previously, there was a `dispatch_strides` boolean arg. Change this to
a string argument that directly maps onto `SizesStridesPolicy`.

suo committed:
Generalize sizes and strides policy on _make_wrapper_subclass

Previously, there was a `dispatch_strides` boolean arg. Change this to
a string argument that directly maps onto `SizesStridesPolicy`.

[ghstack-poisoned]

suo committed:
Update on "Generalize sizes and strides policy on _make_wrapper_subclass"


Previously, there was a `dispatch_strides` boolean arg. Change this to
a string argument that directly maps onto `SizesStridesPolicy`.

[ghstack-poisoned]

albanD commented on torch/csrc/autograd/python_variable.cpp:
```diff
@@ -240,6 +240,22 @@ class PyInterpreterHolder {
+      false,
+      "Unknown sizes_strides_policy: ",
+      arg,
+      "; expected 'strides' or 'sizes'");
```
Is there some doc about what each values does?

suo replied:
I can add documentation for it, where is are the docs for `_make_wrapper_subclass`/`_make_subclass` today?

As for the mismatch—I don't think that's fixable within the scope of this PR (and it already exists). But hopefully we'll finish coverage of the various tensorimpl methods soon and so at least we can document it?

suo replied:
hmmmm, I feel like it's weird to put the rules for what this argument does in the error message; we should probably just have like a real docstring? 

Okay I'm going to merge this now, but I will take as a follow up:
- Finding the right place to describe the rules
- Making the code actually reflect the rules (e.g. make more things overridable from torchdispatch)

albanD reviewed (commented):
Should we have at least one test for the size case?

suo committed:
Update on "Generalize sizes and strides policy on _make_wrapper_subclass"


Previously, there was a `dispatch_strides` boolean arg. Change this to
a string argument that directly maps onto `SizesStridesPolicy`.

[ghstack-poisoned]

suo commented:
@pytorchbot help

suo commented:
@pytorchbot merge -g
